site/content/blog/concourse-ytt.md

```diff
@@ -0,0 +1,395 @@
+
+```yaml
+#@ def task_config():
+platform: linux
+inputs:
+  - name: golang-mock-git
+    path: go/src/github.com/golang/mock
+params:
+  GO111MODULE: "on"
+run:
+  path: /bin/sh
+  args:
+    - -c
+    - |
+      GOPATH=$PWD/go
+
+      cd go/src/github.com/golang/mock
+
+      go vet ./...
+      go build ./...
+      go install github.com/golang/mock/mockgen
+      GO111MODULE=off go get -u golang.org/x/lint/golint
+      ./ci/check_go_fmt.sh
+      ./ci/check_go_lint.sh
+      ./ci/check_go_generate.sh
+      ./ci/check_go_mod.sh
+      go test -v ./...
+#@ end
+
+jobs:
+- name: golang-1.11
+  public: true
+  plan:
+    - get: golang-mock-git
+      trigger: true
+    - get: golang-1.11.x-image
+      trigger: true
+    - task: run-tests
+      image: golang-1.11.x-image
+      config: #@ task_config()
+- name: golang-1.12
+  public: true
+  plan:
+    - get: golang-mock-git
+      trigger: true
+    - get: golang-1.12.x-image
+      trigger: true
+    - task: run-tests
+      image: golang-1.12.x-image
+      config: #@ task_config()
+- name: golang-1.13
+  public: true
+  plan:
+    - get: golang-mock-git
+      trigger: true
+    - get: golang-1.13.x-image
+      trigger: true
+    - task: run-tests
+      image: golang-1.13.x-image
+      config: #@ task_config()
+```
+</details>
+
+   Now that we have most of the jobs and resources refactored, let's draw our attention to all these version numbers. Versions 1.11 to 1.13.
+
+   The first step is to extract these out into a [ytt data value file](https://carvel.dev/ytt/docs/latest/ytt-data-values/#docs). This allows us to have a single place to configurable data.
+
+<details>
+<summary>Values file - values.yml</summary>
+
+```yaml
+#@data/values
+---
+versions: ["1.11", "1.12", "1.13"]
+```
+</details>
+
+   Once we have our data values file, let's go ahead and import the 'data' module so we can use them in our pipeline.yml
+
+<details>
+<summary>Convert to using data.values</summary>
+
+```yaml
+#@ load("@ytt:data", "data")
+
+---
+resources:
+#@ for/end version in data.values.versions:
+- #@ registry_image("golang-" + version + ".x-image", "golang", version + "-stretch")
+
+- name: golang-mock-git
+  type: git
+  icon: github
+  source:
+    uri: https://github.com/golang/mock.git
+
+jobs:
+#@ for/end version in data.values.versions:
+- name: #@ "golang-" + version
+  public: true
+  plan:
+    - get: golang-mock-git
+      trigger: true
+    - get: #@ "golang-" + version + ".x-image"
+      trigger: true
+    - task: run-tests
+      image: #@ "golang-" + version + ".x-image"
+      config: #@ task_config()
+```
+</details>
+
+### Final Pipeline 
+
+Original pipeline 
+
+<details>
+<summary>YTT template for our pipeline</summary>
+
+```yaml
+#@ load("@ytt:data", "data")
+
+#@ def registry_image(name, repository, tag="latest"):
+name: #@ name
+type: registry-image
+icon: docker
+source:
+  repository: #@ repository
+  tag: #@ tag
+#@ end
+
+#@ def task_config():
```

Comment:
```suggestion
#@ def lint_and_test_golang_mock():
```

rationale: name the function so that at the calling site, the reader might understand what's happening without _having_ to read the function body (aka "reveal intention")